src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts

```diff
@@ -585,4 +577,62 @@ export class CoursePlayerComponent implements OnInit, OnDestroy {
     };
     this.telemetryService.interact(telemetryIntractData);
   }
+
+  navigateToPlayerPage(collectionUnit, event?) {
+    if (this.enrolledCourse || this.hasPreviewPermission) {
+      const navigationExtras: NavigationExtras = {
+        queryParams: { batchId: this.batchId, courseId: this.courseId, courseName: this.courseHierarchy.name }
+      };
+      if (event && !_.isEmpty(event.event)) {
+        navigationExtras.queryParams.selectedContent = event.data.identifier;
+      }
+      this.router.navigate(['/learn/course/play', collectionUnit.identifier], navigationExtras);
+    } else {
+      this.showJoinTrainingModal = true;
+    }
+  }
+
+  calculateProgress() {
+    /* istanbul ignore else */
+    if (this.courseHierarchy.children) {
+      this.courseHierarchy.children.forEach(unit => {
+        if (unit.mimeType === 'application/vnd.ekstep.content-collection') {
+          const flattenDeepContents = this.flattenDeep(unit.children).filter(item => item.mimeType !== 'application/vnd.ekstep.content-collection');
+
+          let consumedContents = [];
+          if (consumedContents.length) {
+            consumedContents = flattenDeepContents.filter(o => {
+              return this.contentStatus.some(({ contentId, status }) => o.identifier === contentId && status === 2);
+            });
+          }
+
+          unit.consumedContent = consumedContents.length;
+          unit.contentCount = flattenDeepContents.length;
+          unit.isUnitConsumed = consumedContents.length === flattenDeepContents.length;
+          if (consumedContents.length) {
+            unit.progress = (consumedContents.length / flattenDeepContents.length) * 100;
+          }
+        } else {
+          const consumedContent = this.contentStatus.filter(({ contentId, status }) => unit.identifier === contentId && status === 2);
+          unit.consumedContent = consumedContent.length;
+          unit.contentCount = 1;
+          unit.isUnitConsumed = consumedContent.length === 1;
+        }
+      });
+    }
+  }
+
+  private flattenDeep(contents) {
```

Comment:
This code is there in multiple places suggest to move it to service use it in the components